Codeinwp/otter-blocks
# Add temporary saving for form submissions

Closes #1766 .


This will serve as infrastructure for https://github.com/Codeinwp/otter-blocks/pull/1778

### Summary


Now we can temporarily save a Form submission. When temporary save means:

- validation 
- file upload
- database save

ℹ️ A form request that needs to be saved temporarily is marked by a special HTTP header called `O-Form-Save-Mode` which values must be `temporary`.

ℹ️ Temporary saving is done via a special flag. It is opt-in, which means when someone implements a filter for Form need to consider this case via `is_temporary_data` function of `$form_data`.

The state of this temporary save is shown as `draft` in Form Submissions.

Another API route has been implemented, whose job is to confirm the submission. When confirmed, the form data will be pulled from the database and sent to the processing.

ℹ️ When sent to re-reprocessing, a flag to mark this behavior is set via `is_duplicated` function of `$form_data`. This is also an opt-in. 

For Stripe addon, we want the following workflow:

1. User completes the form and submits it.
2. The data is saved, and a database entry is created.
3. A filter gets the database entry id and use it to create a Stripe session by appending it to the success URL.
4. The stripe session URL is returned.
5. The form gets the response and redirects the client to the Stripe page for payment.
6. After checkout, Stripe redirects the user to the special success URL.
7.  After the page is loaded, the script autodetects the params in URL and call the API endpoint that confirm the submission.

### Screenshots 

### Loading state when confirming the submission.

![image](https://github.com/Codeinwp/otter-blocks/assets/17597852/d77915c7-564c-4718-aab7-e7302c32c69e)


----

### Test instructions


⚠️ You must have `WP_DEBUG` set to `true` and Otter Pro activated.

This test will be done via Console since it is a work of an infrastructure for a future feature.

1. Create a form and preview it.
2. Open Browser Console and write `activateTempSave()`, then press Enter. With this, the form will save the data as temporary.
3. When you submit the form, a new page will open. It will be the same page but with a different URL. Add the end of the URL you will something like `?record_id=123`. Also, you will see a confirmation message like `Success`.

With temporary saving, the final confirmation will be done on the page with the enhanced URL, which means you reached the last step. 



---- 

### Checklist before the final review
